File: ios/Halogen/Features/Downloads/DownloadsModel.swift
import Foundation
import Observation

/// The Downloads tab: the SERVER's download set (Downloaded / Downloading
/// facets). DIVERGENCE from web: the server set is shown for both embedded
/// and remote (the web shows the device set on remote servers).
@MainActor
@Observable
final class DownloadsModel {
    private let accountStore: LocalStore?

    enum Facet: String, CaseIterable, Identifiable {
        case onDevice = "on_device"
        case downloaded
        case downloading

        var id: String { rawValue }
        var label: String {
            switch self {
            case .onDevice: return "On device"
            case .downloaded: return "Server"
            case .downloading: return "Downloading"
            }
        }
        var token: String { rawValue.uppercased() }
    }

    private unowned let core: HalogenCore

    private(set) var episodes: [EpisodeData] = []
    private(set) var loaded = false
    private(set) var error: String?
    private(set) var hasMore = false
    /// Last loadMore failed — the sentinel renders a retry (LoadMoreRow).
    private(set) var loadMoreFailed = false
    private var page = 0
    private var loadingMore = false
    var facet: Facet = .onDevice
    /// Search + order (facet fixed by the segmented control). Persisted
    /// across relaunch (web use_list_view_state("downloads")).
    var query = ListQuery() {
        didSet {
            guard query != oldValue else { return }
            let snapshot = query
            Task { [store = accountStore] in
                await store?.save(snapshot, key: "listquery-downloads")
            }
        }
    }
    private var loadedQuery = false

    /// Embedded accounts have no device tier — the server set IS this device.
    var availableFacets: [Facet] {
        core.isEmbeddedAccount ? [.downloaded, .downloading] : Facet.allCases
    }

    init(core: HalogenCore) {
        self.core = core
        self.accountStore = core.store
        if core.isEmbeddedAccount {
            facet = .downloaded
        }
    }

    /// Unsubscribe cascade: drop the podcast's rows from the visible list
    /// (the snapshots are pruned by the core's purge).
    func removePodcastLocally(_ podcastId: Int32) {
        episodes.removeAll { $0.podcast_id == podcastId }
    }

    /// The facet whose rows currently populate `episodes` — a facet switch
    /// must repaint from ITS cache, but a mere query keystroke must not.
    private var paintedFacet: Facet?

    func load() async {
        error = nil
        if !loadedQuery {
            loadedQuery = true
            if let store = accountStore,
                let saved = await store.load(ListQuery.self, key: "listquery-downloads")
            {
                query = saved
            }
        }
        // Cache-paint only a COLD list or a facet switch (siblings' rule):
        // this refires on every query keystroke, and repainting the full
        // unfiltered snapshot over live search results flashes the whole
        // list per key. `loaded` stays as-is (no empty-state blink).
        let facetChanged = paintedFacet != facet
        if episodes.isEmpty || facetChanged, query.search.isEmpty {
            if facetChanged { episodes = [] }
            if let store = accountStore,
                let cached = await store.load(
                    [EpisodeData].self, key: CacheKey.downloads(facet.rawValue))
            {
                episodes = cached
                loaded = true
            }
        }
        paintedFacet = facet
        await refresh()
    }

    private var generation = 0

    func refresh() async {
        if !query.search.isEmpty {
            do { try await Task.sleep(for: .milliseconds(300)) } catch { return }
        }
        if facet == .onDevice {
            // The device set is local — query applies in memory, fully offline.
            let all = core.models?.device.onDevice ?? []
            episodes = query.apply(
                to: all,
                status: { [weak core] in core?.overlayStatus($0) ?? $0.playback_status ?? .unplayed })
            error = nil
            loaded = true
            return
        }
        do {
            let first = try await core.forAccount(accountStore).latestEpisodesRaw(
                extra: [URLQueryItem(name: "filter[download_status]", value: facet.token)]
                    + query.queryItems,
                page: 0, pageSize: 20)
            episodes = first.items
            hasMore = first.hasMore
            page = 0
            error = nil
            if query.search.isEmpty {
                // Prefix-merge like the other list snapshots: fresh page 0
                // leads, the cached tail survives so scrolled-through rows
                // stay renderable offline.
                let ids = Set(first.items.map(\.id))
                var snapshot = first.items
                if let prior = await accountStore?.load(
                    [EpisodeData].self, key: CacheKey.downloads(facet.rawValue))
                {
                    snapshot += prior.filter { !ids.contains($0.id) }
                }
                await accountStore?.save(snapshot, key: CacheKey.downloads(facet.rawValue))
            }
        } catch {
            if episodes.isEmpty { self.error = FriendlyError.message(error) }
        }
        loaded = true
    }

    func loadMore() async {
        guard facet != .onDevice, hasMore, !loadingMore, loaded else { return }
        loadingMore = true
        defer { loadingMore = false }
        loadMoreFailed = false
        do {
            let next = try await core.forAccount(accountStore).latestEpisodesRaw(
                extra: [URLQueryItem(name: "filter[download_status]", value: facet.token)]
                    + query.queryItems,
                page: page + 1, pageSize: 20)
            page += 1
            let known = Set(episodes.map(\.id))
            episodes.append(contentsOf: next.items.filter { !known.contains($0.id) })
            hasMore = next.hasMore
        } catch {
            loadMoreFailed = true
        }
    }

    /// On-device facet: delete the local copy. Server facets: delete the
    /// server file — durable (web: RemoveServerDownload op), so it queues
    /// offline instead of silently failing.
    func removeDownload(_ episode: EpisodeData) async {
        if facet == .onDevice {
            episodes.removeAll { $0.id == episode.id }
            core.models?.device.remove(episode.id)
            return
        }
        guard await core.ensureQueued(originStore: accountStore, .removeServerDownload(episodeId: episode.id)) else {
            return
        }
        episodes.removeAll { $0.id == episode.id }
        // Optimistic overlay (every row/menu flips immediately) AND snapshot
        // patch — load()'s cache-paint must not resurrect the removed row.
        core.models?.serverDownloads.markRemovedLocally(episode.id)
        let snapshot = episodes
        await accountStore?.save(snapshot, key: CacheKey.downloads(facet.rawValue))
    }
}

File: android/app/src/main/kotlin/org/fgsec/halogen/features/downloads/DownloadsModel.kt
package org.fgsec.halogen.features.downloads

import org.fgsec.halogen.core.ensureQueued
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.fgsec.halogen.components.ListQuery
import org.fgsec.halogen.core.DeviceLog
import org.fgsec.halogen.core.FriendlyError
import org.fgsec.halogen.core.HalogenCore
import org.fgsec.halogen.storage.CacheKey
import org.fgsec.halogen.storage.OutboxOp
import org.fgsec.halogen.wire.EpisodeData

/// The Downloads tab: On-device (this device's local copies, fully offline)
/// plus the SERVER's download set (Server / Downloading facets). Embedded
/// accounts have no device tier — the server set IS this device.
class DownloadsModel(private val core: HalogenCore) {
    private val accountStore = core.store


    enum class Facet(val rawValue: String) {
        OnDevice("on_device"),
        Downloaded("downloaded"),
        Downloading("downloading");

        val label: String
            get() = when (this) {
                OnDevice -> "On device"
                Downloaded -> "Server"
                Downloading -> "Downloading"
            }

        /// The wire `filter[download_status]` token (server facets only).
        val token: String get() = rawValue.uppercase()
    }

    var episodes: List<EpisodeData> by mutableStateOf(emptyList())
        private set
    var loaded: Boolean by mutableStateOf(false)
        private set
    var error: String? by mutableStateOf(null)
        private set
    var hasMore: Boolean by mutableStateOf(false)
        private set
    /// Last loadMore failed — the sentinel renders a retry (LoadMoreRow).
    var loadMoreFailed: Boolean by mutableStateOf(false)
        private set
    private var page = 0
    private var loadingMore = false
    var facet: Facet by mutableStateOf(
        if (core.isEmbeddedAccount) Facet.Downloaded else Facet.OnDevice)

    /// Search + order (facet fixed by the segmented control). Persisted
    /// across relaunch (web use_list_view_state("downloads")).
    private var queryState: ListQuery by mutableStateOf(ListQuery())
    var query: ListQuery
        get() = queryState
        set(value) {
            if (value == queryState) return
            queryState = value
            val snapshot = value
            core.scope.launch { accountStore?.save(snapshot, "listquery-downloads") }
        }
    private var loadedQuery = false

    /// Embedded accounts have no device tier — the server set IS this device.
    val availableFacets: List<Facet>
        get() = if (core.isEmbeddedAccount) listOf(Facet.Downloaded, Facet.Downloading)
        else Facet.entries.toList()

    /// Unsubscribe cascade: drop the podcast's rows from the visible list
    /// (the snapshots are pruned by the core's purge).
    fun removePodcastLocally(podcastId: Int) {
        episodes = episodes.filterNot { it.podcast_id == podcastId }
    }

    /// The facet whose rows currently populate `episodes` — a facet switch
    /// must repaint from ITS cache, but a mere query keystroke must not.
    private var paintedFacet: Facet? = null

    suspend fun load() {
        error = null
        if (!loadedQuery) {
            loadedQuery = true
            accountStore?.load<ListQuery>("listquery-downloads")?.let { query = it }
        }
        // Cache-paint only a COLD list or a facet switch (siblings' rule):
        // this refires on every query keystroke, and repainting the full
        // unfiltered snapshot over live search results flashes the whole
        // list per key. `loaded` stays as-is (no empty-state blink).
        val facetChanged = paintedFacet != facet
        if ((episodes.isEmpty() || facetChanged) && query.search.isEmpty()) {
            if (facetChanged) episodes = emptyList()
            accountStore?.load<List<EpisodeData>>(CacheKey.downloads(facet.rawValue))?.let { cached ->
                episodes = cached
                loaded = true
            }
        }
        paintedFacet = facet
        refresh()
    }

    suspend fun refresh() {
        // Debounce typing (the view's keyed effect cancels the sleep per keystroke).
        if (query.search.isNotEmpty()) delay(300)
        if (facet == Facet.OnDevice) {
            // The device set is local — query applies in memory, fully offline.
            val all = core.models?.device?.onDevice ?: emptyList()
            episodes = query.apply(all, status = { core.overlayStatus(it) })
            error = null
            loaded = true
            return
        }
        try {
            val first = core.latestEpisodesRaw(
                extra = listOf("filter[download_status]" to facet.token) + query.queryItems,
                page = 0, pageSize = 20)
            episodes = first.items
            hasMore = first.hasMore
            page = 0
            error = null
            if (query.search.isEmpty()) {
                // Prefix-merge like the other list snapshots: fresh page 0
                // leads, the cached tail survives so scrolled-through rows
                // stay renderable offline.
                val ids = first.items.map { it.id }.toSet()
                var snapshot = first.items
                accountStore?.load<List<EpisodeData>>(CacheKey.downloads(facet.rawValue))
                    ?.let { prior -> snapshot = snapshot + prior.filterNot { it.id in ids } }
                accountStore?.save(snapshot, CacheKey.downloads(facet.rawValue))
            }
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            DeviceLog.warn("downloads: refresh failed — ${e::class.simpleName}: ${e.message}")
            if (episodes.isEmpty()) error = FriendlyError.message(e)
        }
        loaded = true
    }

    suspend fun loadMore() {
        if (facet == Facet.OnDevice || !hasMore || loadingMore || !loaded) return
        loadingMore = true
        loadMoreFailed = false
        try {
            val next = core.latestEpisodesRaw(
                extra = listOf("filter[download_status]" to facet.token) + query.queryItems,
                page = page + 1, pageSize = 20)
            page += 1
            val known = episodes.map { it.id }.toSet()
            episodes = episodes + next.items.filterNot { it.id in known }
            hasMore = next.hasMore
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            DeviceLog.warn("downloads: loadMore failed — ${e::class.simpleName}: ${e.message}")
            loadMoreFailed = true
        } finally {
            loadingMore = false
        }
    }

    /// On-device facet: delete the local copy. Server facets: delete the
    /// server file — durable (web: RemoveServerDownload op), so it queues
    /// offline instead of silently failing.
    suspend fun removeDownload(episode: EpisodeData) {
        if (facet == Facet.OnDevice) {
            episodes = episodes.filterNot { it.id == episode.id }
            core.models?.device?.remove(episode.id)
            return
        }
        if (!core.ensureQueued(OutboxOp.Kind.RemoveServerDownload(episodeId = episode.id))) return
        episodes = episodes.filterNot { it.id == episode.id }
        core.models?.serverDownloads?.markRemovedLocally(episode.id)
    }
}
